### ISSUE TITLE
hotflip api not working for some models

### ISSUE THREAD

**Issue by aimichal** [allenai/allennlp#4355]
Example request to NAQANET resulting in HTTP 500:

```
curl -sv 'https://demo.allennlp.org/api/naqanet/attack/hotflip' -H 'Content-Type: application/json' --data-raw '{"inputs":{"model":"naqanet","passage":"Robotics is an interdisciplinary branch of engineering and science that includes mechanical engineering, electrical engineering, computer science, and others. Robotics deals with the design, construction, operation, and use of robots, as well as computer systems for their control, sensory feedback, and information processing. These technologies are used to develop machines that can substitute for humans. Robots can be used in any situation and for any purpose, but today many are used in dangerous environments (including bomb detection and de-activation), manufacturing processes, or where humans cannot survive. Robots can take on any form but some are made to resemble humans in appearance. This is said to help in the acceptance of a robot in certain replicative behaviors usually performed by people. Such robots attempt to replicate walking, lifting, speech, cognition, and basically anything a human can do.","question":"What do robots that resemble humans attempt to do?"},"input_field_to_attack":"question","grad_input_field":"grad_input_2"}'
```

The logs show this stack trace:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/app/allennlp_demo/common/http.py", line 210, in attack_handler
    with_cache_hit_response_headers(self.attack_with_cache, attacker_id, request.data)
  File "/app/allennlp_demo/common/http.py", line 40, in with_cache_hit_response_headers
    r = fn(*args)
  File "/app/allennlp_demo/common/http.py", line 103, in attack_with_cache
    return self.attack(attacker_id, json.loads(attack))
  File "/app/allennlp_demo/common/http.py", line 163, in attack
    return attacker.attack_from_json(**attack)
  File "/usr/local/lib/python3.6/site-packages/allennlp/interpret/attackers/hotflip.py", line 280, in attack_from_json
    grad[index_of_token_to_flip], original_id_of_token_to_flip, sign
  File "/usr/local/lib/python3.6/site-packages/allennlp/interpret/attackers/hotflip.py", line 334, in _first_order_taylor
    inputs = self._make_embedder_input([self.vocab.get_token_from_index(token_idx)])
  File "/usr/local/lib/python3.6/site-packages/allennlp/data/vocabulary.py", line 665, in get_token_from_index
    return self._index_to_token[namespace][index]
KeyError: tensor(28894)
```

And with bidaf-elmo:

```
curl -sv 'https://demo.allennlp.org/api/bidaf-elmo/attack/hotflip' -H 'Content-Type: application/json' --data-raw '{"inputs":{"passage":"A reusable launch system (RLS, or reusable launch vehicle, RLV) is a launch system which is capable of launching a payload into space more than once. This contrasts with expendable launch systems, where each launch vehicle is launched once and then discarded. No completely reusable orbital launch system has ever been created. Two partially reusable launch systems were developed, the Space Shuttle and Falcon 9. The Space Shuttle was partially reusable: the orbiter (which included the Space Shuttle main engines and the Orbital Maneuvering System engines), and the two solid rocket boosters were reused after several months of refitting work for each launch. The external tank was discarded after each flight.","question":"How many partially reusable launch systems were developed?"},"input_field_to_attack":"question","grad_input_field":"grad_input_2"}'
```

The logs show this stack trace:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/app/allennlp_demo/common/http.py", line 210, in attack_handler
    with_cache_hit_response_headers(self.attack_with_cache, attacker_id, request.data)
  File "/app/allennlp_demo/common/http.py", line 40, in with_cache_hit_response_headers
    r = fn(*args)
  File "/app/allennlp_demo/common/http.py", line 103, in attack_with_cache
    return self.attack(attacker_id, json.loads(attack))
  File "/app/allennlp_demo/common/http.py", line 163, in attack
    return attacker.attack_from_json(**attack)
  File "/usr/local/lib/python3.6/site-packages/allennlp/interpret/attackers/hotflip.py", line 280, in attack_from_json
    grad[index_of_token_to_flip], original_id_of_token_to_flip, sign
  File "/usr/local/lib/python3.6/site-packages/allennlp/interpret/attackers/hotflip.py", line 330, in _first_order_taylor
    if token_idx >= self.embedding_matrix.size(0):
RuntimeError: bool value of Tensor with more than one value is ambiguous
```

**Comment by matt-gardner** [comment]
@aimichal, I'd say that the UI should have some graceful way of handling 500s for various interpret results, because they tend to be less stable.  If they're not handled nicely in UI, that's a demo bug, but the underlying cause of the 500 is an allennlp bug.

**Comment by matt-gardner** [comment]
Thanks for finding these, also.  I will look into them (at some point...).

**Comment by aimichal** [comment]
> graceful way of handling 500s

I agree, and wanted to make that improvement quickly initially, at least, but the code was hard to follow so I stopped. I'll file a thing to make the UI more robust: https://github.com/allenai/allennlp-demo/issues/489

**Comment by matt-gardner** [comment]
I'm adding this to the 1.2 milestone so we don't forget about it, but it really belongs in 1.3 or 2.1 or something, with a target fix date of end of October.
